test/e2e/instrumentation/tracing/apiserver.go

```diff
@@ -0,0 +1,190 @@
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/component-base/traces"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+	e2eskipper "k8s.io/kubernetes/test/e2e/framework/skipper"
+	e2essh "k8s.io/kubernetes/test/e2e/framework/ssh"
+	instrumentation "k8s.io/kubernetes/test/e2e/instrumentation/common"
+)
+
+const (
+	podName       = "otel-collector"
+	containerName = "collector"
+)
+
+// The API Server Tracing test ensures that an opentelemetry collector can
+// collect traces from the API Server, and that context is correctly propagated
+var _ = instrumentation.SIGDescribe("[Feature:APIServerTracing]", func() {
+	f := framework.NewDefaultFramework("apiserver-tracing")
+	var c clientset.Interface
+	var otelPod *v1.Pod
+
+	ginkgo.BeforeEach(func() {
+		config, err := framework.LoadConfig()
+		framework.ExpectNoError(err)
+		// Use otelhttp with the client-go client to ensure context is propagated.
+		config.Wrap(func(rt http.RoundTripper) http.RoundTripper {
+			return otelhttp.NewTransport(rt)
+		})
+		c, err = clientset.NewForConfig(config)
+		framework.ExpectNoError(err)
+
+		ginkgo.By("Creating an opentelemetry collector on the master node, which logs spans to stdout.")
+		masterNode, err := masterNodeName(c)
+		framework.ExpectNoError(err)
+		otelPod, err = c.CoreV1().Pods(f.Namespace.Name).Create(context.TODO(), opentelemetryCollectorPod(masterNode), metav1.CreateOptions{})
+		framework.ExpectNoError(err)
+
+		_, err = c.CoreV1().ConfigMaps(f.Namespace.Name).Create(context.TODO(), opentelemetryConfigmap(), metav1.CreateOptions{})
+		framework.ExpectNoError(err)
+
+		ready := e2epod.CheckPodsRunningReady(f.ClientSet, f.Namespace.Name, []string{podName}, 20*time.Second)
+		framework.ExpectEqual(ready, true)
+	})
+
+	ginkgo.It("should send a request with a sampled trace context, and observe child spans from the collector pod", func() {
+		e2eskipper.SkipUnlessSSHKeyPresent()
+
+		ginkgo.By("Setting up OpenTelemetry")
+		tp := sdktrace.NewTracerProvider()
+		otel.SetTracerProvider(tp)
+		otel.SetTextMapPropagator(traces.Propagators())
+
+		ginkgo.By("Creating a context with a  parent span")
+		ctx, span := tp.Tracer("apiservertest").Start(context.Background(), "OpenTelemetrySpan")
+
+		traceID := span.SpanContext().TraceID()
+		ginkgo.By(fmt.Sprintf("Checking for Trace ID: %v in logs.", traceID.String()))
+		gomega.Eventually(func() error {
+			// Send any request using the context with a sampled span
+			_, err := c.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+			if err != nil {
+				return err
+			}
+
+			// Get logs from the opentelemetry collector pod on the master.
+			// We must use SSH because we can't fetch logs from master pods
+			// using the pod logs subresource.
+			result, err := e2essh.SSH(
+				fmt.Sprintf("sudo cat /var/log/pods/%v_%v_%v/%v/*", f.Namespace.Name, podName, otelPod.UID, containerName),
+				framework.APIAddress()+":22",
+				framework.TestContext.Provider,
+			)
```

Comment:
Is this really the best way to check? There isn't another way?

Reply:
Here is my current idea: We can use the [file](https://github.com/open-telemetry/opentelemetry-collector/tree/main/exporter/fileexporter#file-exporter) exporter on the collector, which writes traces to a file in json format.  We can run a busybox sidecar container that periodically `cat`s the file, greps for whatever we want to match, and exits with code 0 if it finds what its looking for.

We can run one pod on each master node we find (instead of just the first one we find).  The test waits for one of the sidecar containers to exit with exit code 0, or times out.

WDYT?

Reply:
it could be better but more moving pieces so possibly flakier? it's hard to say..